### PR TITLE
Oauth2 extensions

### PR DESCRIPTION
**Pull request by kevprice83**
Initial logic for support for all flows and basic test coverage

### CODE REVIEW

File: apicast/conf/nginx.conf [hunk 1]
```diff
@@ -4,6 +4,7 @@ env REDIS_URL;
 env RESOLVER;
 env BACKEND_ENDPOINT_OVERRIDE;
 env RHSSO_ENDPOINT;
+env APICAST_OAUTH_ACCESS_TOKEN_TTL;
```
Comment (mikz) [inline apicast/conf/nginx.conf]: All variables starting with `APICAST_` are forwarded automatically. This is not necessary.

File: apicast/src/oauth/apicast_oauth.lua [hunk 2]
```diff
@@ -1,26 +1,417 @@
...
+
+-- returns a unique string for the client_id. it will be short lived
+local function nonce(client_id)
+  return ts.sha1_digest(tostring(random.bytes(20, true)) .. "#login:" .. client_id)
+end
+
+local function generate_access_token(client_id)
+  local token = ts.sha1_digest(tostring(random.bytes(20, true)) .. client_id)
+
+  return { ["access_token"] = token, ["token_type"] = "bearer", ["expires_in"] = env.get('APICAST_OAUTH_ACCESS_TOKEN_TTL') or 604800 }
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: no need to use string keys, those are all valid lua table identifiers without quotes

Comment (kevprice83) [inline apicast/src/oauth/apicast_oauth.lua]: Okay noted, will change

File: apicast/src/oauth/apicast_oauth.lua [hunk 3]
```diff
@@ -1,26 +1,432 @@
...
+
+function _M.respond_with_error(status, message, description)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message, error_description = description }
+  local body = cjson.encode(err_msg)
+  ngx.log(ngx.INFO, "error :" .. inspect(body))
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: no need to inspect a string returned from cjson.encode, also never ever concatenate strings when logging, pass it as parameters

File: apicast/src/oauth/apicast_oauth.lua [hunk 4]
```diff
@@ -1,26 +1,432 @@
...
+  _M.respond_and_exit(status, body, headers)
+end
+
+-- TODO: Split error conditions up further to decide when we should respond with error and when we should redirect_with error
+function _M.redirect_with_error(url, error, state)
+  ngx.header.content_type = "application/x-www-form-urlencoded"
+  return ngx.redirect(url,"?error=",error.error,"&error_description=",error.error_description,"&state=",state)
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: Just use `ngx.encode_args` to properly serialize the query string

File: apicast/src/oauth/apicast_oauth.lua [hunk 5]
```diff
@@ -1,26 +1,432 @@
...
+local function retrieve_client_data(service_id, params)
+  local tmp_data = service_id .. "#tmp_data:".. params.state
+  local client_data
+  local redis = ts.connect_redis()
+
+  if redis then
+    local ok, err = redis:hgetall(tmp_data)
+
+    if not ok then
+      ngx.log(0, "no values for tmp_data hash: ".. ts.dump(err))
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: no `ngx.log(0` please, and no concatenating strings, and no calling functions when passing values to log, just pass the plain err

File: apicast/src/oauth/apicast_oauth.lua [hunk 6]
```diff
@@ -1,26 +1,432 @@
...
+      ngx.header.content_type = "application/x-www-form-urlencoded"
+      return ngx.redirect(params.redirect_uri .. "#error=invalid_request&error_description=invalid_or_expired_state&state=" .. (params.state or ""))
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: This does not release redis.

File: apicast/src/oauth/apicast_oauth.lua [hunk 7]
```diff
@@ -1,26 +1,432 @@
...
+    ts.release_redis(redis)
+  end
+
+  return client_data
+end
+
+-- Returns the code to the client
+local function send_code(client_data, code)
+  ngx.header.content_type = "application/x-www-form-urlencoded"
+  return ngx.redirect( client_data.redirect_uri .. "?code="..code.."&state=" .. (client_data.state or ""))
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: please prefer `ngx.encode_args` over creating them by hand. they should be escaped properly

File: apicast/src/oauth/apicast_oauth.lua [hunk 8]
```diff
@@ -1,26 +1,432 @@
...
+  if redis then
+    ok, err = redis:hmset("c:".. code, {
+      client_id = client_data.client_id,
+      client_secret = client_data.secret_id,
+      redirect_uri = client_data.redirect_uri,
+      code = code
+    })
+
+    if ok then
+      return redis:expire("c:".. code, 60 * 10) -- code expires in 10 mins
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: This never releases redis.

File: apicast/src/oauth/apicast_oauth.lua [hunk 9]
```diff
@@ -1,26 +1,432 @@
...
+  local err
+  local access_token = generate_access_token(params.client_id)
+
+  local stored = store_token(params, access_token)
+
+  if stored.status == 200 then
+    send_token(access_token)
+    return
+  else
+    err =  '{"error":"'..stored.body..'"}'
```
Comment (mikz) [inline apicast/src/oauth/apicast_oauth.lua]: `cjson.encode`

**Comment by mikz** [comment]
No need to change my comments from the review. Just for the next time.

File: apicast/src/oauth/apicast_oauth.lua (hunk 3 above)
Comment (kevprice83) [inline apicast/src/oauth/apicast_oauth.lua]: ah this was just some logging I forgot to remove. I'll clean up, thanks. Probably should add some more useful logging though...

File: apicast/src/oauth/apicast_oauth.lua (hunk 5 above)
Comment (kevprice83) [inline apicast/src/oauth/apicast_oauth.lua]: right this is probably resource intensive when logging like this...I will look for any other occurrences of logging like this

**Comment by kevprice83** [comment]
@vramosp what should be done with this PR? This was requested by product for end of April and provided. Are we going to maybe maintain this as a separate branch or just discontinue completely? The new OAuth integration will support some but not all flows cc @mikz and the integration would be with an external Auth server not with APIcast acting as the Auth server which is the case in this PR.

**Comment by MarkCheshire** [comment]
The decision is that the final implementation in the upstream and product may differ from this code. Therefore this will remain as an unmerged branch. This functionality can be used to implement OAuth2 flows at any time. An important prerequisite is to rebase this branch on the latest production release of APIcast (check the release tags).

**Comment by mikz** [comment]
Closing this PR. The work does not continue and it is not indented to be merged. Reopen if you feel differently.

**Pull request closed without merging**